### PR TITLE
8257772: Vectorizing clear memory operation using AVX-512 masked operations

### PR DESCRIPTION
**Pull request by jatin-bhateja**
A newly allocated memory is initialized either using user provided initialization values for various fields or setting the memory to zero as per java semantics (System initialization).

C2 compiler creates ClearArray Node in order to perform system initialization. ClearArray accepts the number of Heap Words to be initialized, this number can be constant or a non-constant value. For constant number of heap words less than InitArrayShortSize (default value 64 bytes) currently compiler generates StoreL nodes which does the initialization at the granularity of 8 bytes.

This patch vectorizes the initializing store operations for constant sized heap word less than InitArrayShortSize by emitting special instruction sequence for various tail sizes.

In addition existing implementation for initialization under UseXMMForObjInit is extended to use masked operation to optimize tail initialization sequence. In case AVX3Threshold is set to 0 then new initialization sequence uses 64 byte ZMM registers.

Following are the performance stats collected using  micro-benchmark included with the patch.

Testing : Tier1-Tier3 level tests are clean.

System Configuration : Cascadelake, Intel Xeon Platinum 8280L @ 2.7 GHz, 2 socket, 28 cores per socket.

### Baseline:
```
Benchmark                        Mode  Cnt          Score   Error  Units
ClearMemory.testClearMemory16K  thrpt    2    1427741.069          ops/s
ClearMemory.testClearMemory1K   thrpt    2   47628368.596          ops/s
ClearMemory.testClearMemory1M   thrpt    2      27388.979          ops/s
ClearMemory.testClearMemory24B  thrpt    2  167681010.419          ops/s
ClearMemory.testClearMemory2K   thrpt    2   22043948.290          ops/s
ClearMemory.testClearMemory32B  thrpt    2  168599498.817          ops/s
ClearMemory.testClearMemory32K  thrpt    2     775985.067          ops/s
ClearMemory.testClearMemory40B  thrpt    2  153375273.800          ops/s
ClearMemory.testClearMemory48B  thrpt    2  145328531.804          ops/s
ClearMemory.testClearMemory4K   thrpt    2    6492257.452          ops/s
ClearMemory.testClearMemory56B  thrpt    2  122376321.652          ops/s
ClearMemory.testClearMemory8K   thrpt    2    2857444.413          ops/s
ClearMemory.testClearMemory8M   thrpt    2       3461.674          ops/s
```
### With Optimization:
```
Benchmark                        Mode  Cnt          Score   Error  Units
ClearMemory.testClearMemory16K  thrpt    2    2529701.368          ops/s
ClearMemory.testClearMemory1K   thrpt    2   50276682.550          ops/s
ClearMemory.testClearMemory1M   thrpt    2      27458.588          ops/s
ClearMemory.testClearMemory24B  thrpt    2  178751174.642          ops/s
ClearMemory.testClearMemory2K   thrpt    2   22574802.694          ops/s
ClearMemory.testClearMemory32B  thrpt    2  176630844.950          ops/s
ClearMemory.testClearMemory32K  thrpt    2    1297627.181          ops/s
ClearMemory.testClearMemory40B  thrpt    2  167469550.653          ops/s
ClearMemory.testClearMemory48B  thrpt    2  159391163.006          ops/s
ClearMemory.testClearMemory4K   thrpt    2    9045158.643          ops/s
ClearMemory.testClearMemory56B  thrpt    2  134550172.421          ops/s
ClearMemory.testClearMemory8K   thrpt    2    4581450.664          ops/s
ClearMemory.testClearMemory8M   thrpt    2       3446.834          ops/s
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8257772](https://bugs.openjdk.java.net/browse/JDK-8257772): Vectorizing clear memory operation using AVX-512 masked operations

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1631/head:pull/1631`
`$ git checkout pull/1631`

### CODE REVIEW

File: src/hotspot/cpu/x86/x86_64.ad [hunk 1]
```diff
@@ -10853,6 +10853,18 @@ instruct rep_stos_large(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
   ins_pipe(pipe_slow);
 %}
 
+instruct rep_stos_im(immL cnt, rRegP base, regD tmp, rRegI zero, Universe dummy, rFlagsReg cr)
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86_64.ad:10856]: What about x86_32.ad?

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86_64.ad:10856]: Hi Vladimir, Pattern has been added in x86_32.ad

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Submitted some quick testing for this and there are failures with tests in `compiler/c2/cr6340864/`:
```
#  Internal Error (workspace/open/src/hotspot/cpu/x86/macroAssembler_x86.cpp:8178), pid=27510, tid=27529
#  assert(MaxVectorSize >= 32) failed: vector length should be >= 32

Current CompileTask:
C2:    259   28    b        java.lang.StringCoding::encodeASCII (158 bytes)

Stack: [0x00007f2d144f8000,0x00007f2d145f9000],  sp=0x00007f2d145f3750,  free space=1005k
Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x13a326c]  MacroAssembler::fill64_avx(RegisterImpl*, int, XMMRegisterImpl*, bool)+0x11c
V  [libjvm.so+0x13a3415]  MacroAssembler::xmm_clear_mem(RegisterImpl*, RegisterImpl*, RegisterImpl*, XMMRegisterImpl*)+0x195
V  [libjvm.so+0x13a458b]  MacroAssembler::clear_mem(RegisterImpl*, RegisterImpl*, RegisterImpl*, XMMRegisterImpl*, bool)+0x19b
V  [libjvm.so+0x395487]  rep_stosNode::emit(CodeBuffer&, PhaseRegAlloc*) const+0x167
V  [libjvm.so+0x15b79da]  PhaseOutput::scratch_emit_size(Node const*)+0x3fa
V  [libjvm.so+0x15ae88c]  PhaseOutput::shorten_branches(unsigned int*)+0x2ac
V  [libjvm.so+0x15c045a]  PhaseOutput::Output()+0xcda
V  [libjvm.so+0xa0a798]  Compile::Code_Gen()+0x438
V  [libjvm.so+0xa13fe7]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0x1917
V  [libjvm.so+0x8466ac]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1dc
V  [libjvm.so+0xa24498]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xe08
V  [libjvm.so+0xa24fe8]  CompileBroker::compiler_thread_loop()+0x5a8
V  [libjvm.so+0x18ae756]  JavaThread::thread_main_inner()+0x256
V  [libjvm.so+0x18b50e0]  Thread::call_run()+0x100
V  [libjvm.so+0x1598346]  thread_native_entry(Thread*)+0x116
```

Tests are executed with `-XX:CompileThreshold=100 -XX:-TieredCompilation`.

**Comment by jatin-bhateja** [comment]
> Submitted some quick testing for this and there are failures with tests in `compiler/c2/cr6340864/`:
> 
> ```
> #  Internal Error (workspace/open/src/hotspot/cpu/x86/macroAssembler_x86.cpp:8178), pid=27510, tid=27529
> #  assert(MaxVectorSize >= 32) failed: vector length should be >= 32
> 
> Current CompileTask:
> C2:    259   28    b        java.lang.StringCoding::encodeASCII (158 bytes)
> 
> Stack: [0x00007f2d144f8000,0x00007f2d145f9000],  sp=0x00007f2d145f3750,  free space=1005k
> Native frames: (J=compiled Java code, A=aot compiled Java code, j=interpreted, Vv=VM code, C=native code)
> V  [libjvm.so+0x13a326c]  MacroAssembler::fill64_avx(RegisterImpl*, int, XMMRegisterImpl*, bool)+0x11c
> V  [libjvm.so+0x13a3415]  MacroAssembler::xmm_clear_mem(RegisterImpl*, RegisterImpl*, RegisterImpl*, XMMRegisterImpl*)+0x195
> V  [libjvm.so+0x13a458b]  MacroAssembler::clear_mem(RegisterImpl*, RegisterImpl*, RegisterImpl*, XMMRegisterImpl*, bool)+0x19b
> V  [libjvm.so+0x395487]  rep_stosNode::emit(CodeBuffer&, PhaseRegAlloc*) const+0x167
> V  [libjvm.so+0x15b79da]  PhaseOutput::scratch_emit_size(Node const*)+0x3fa
> V  [libjvm.so+0x15ae88c]  PhaseOutput::shorten_branches(unsigned int*)+0x2ac
> V  [libjvm.so+0x15c045a]  PhaseOutput::Output()+0xcda
> V  [libjvm.so+0xa0a798]  Compile::Code_Gen()+0x438
> V  [libjvm.so+0xa13fe7]  Compile::Compile(ciEnv*, ciMethod*, int, bool, bool, bool, bool, DirectiveSet*)+0x1917
> V  [libjvm.so+0x8466ac]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x1dc
> V  [libjvm.so+0xa24498]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xe08
> V  [libjvm.so+0xa24fe8]  CompileBroker::compiler_thread_loop()+0x5a8
> V  [libjvm.so+0x18ae756]  JavaThread::thread_main_inner()+0x256
> V  [libjvm.so+0x18b50e0]  Thread::call_run()+0x100
> V  [libjvm.so+0x1598346]  thread_native_entry(Thread*)+0x116
> ```
> 
> Tests are executed with `-XX:CompileThreshold=100 -XX:-TieredCompilation`.

Hi Tobi, thanks,  I missed a safety check for MaxVectorSize >= 32 in xmm_clear_mem, for platforms supporting AVX feature, I have fixed this and running tests,  can you kindly run the patch with default options over your internal performance suite and confirm if there is not performance degradation.

**Comment by TobiHartmann** [comment]
Okay, will do and report back once it finished.

**Comment by TobiHartmann** [comment]
Just noticed that you didn't update the patch yet. Could you first push the fix?

**Comment by jatin-bhateja** [comment]
> Just noticed that you didn't update the patch yet. Could you first push the fix?

Done

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 2]
```diff
@@ -4927,22 +4927,26 @@ void MacroAssembler::verified_entry(int framesize, int stack_bang_size, bool fp_
...
     pxor(xtmp, xtmp);
   }
   jmp(L_zero_64_bytes);
 
   BIND(L_loop);
-  if (UseAVX >= 2) {
-    vmovdqu(Address(base,  0), xtmp);
-    vmovdqu(Address(base, 32), xtmp);
+  if (MaxVectorSize >= 32) {
+    fill64_avx(base,  0, xtmp, use64byteVector);
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:4949]: Extra whitespace after `base,  `

File: test/micro/org/openjdk/bench/vm/compiler/ClearMemory.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+import java.util.concurrent.TimeUnit;
+
+@Fork(jvmArgsPrepend = {"-XX:-EliminateAllocations", "-XX:-DoEscapeAnalysis"})
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.SECONDS)
+@State(Scope.Thread)
+public class ClearMemory {
+  class Payload8 {
+    public long f0;
```
Comment (TobiHartmann) [inline test/micro/org/openjdk/bench/vm/compiler/ClearMemory.java:45]: For Java code we use 4 whitespace indentation.

File: src/hotspot/cpu/x86/x86_64.ad [hunk 4]
```diff
@@ -10853,6 +10853,18 @@ instruct rep_stos_large(rcx_RegL cnt, rdi_RegP base, regD tmp, rax_RegI zero,
...
+%{
+  predicate(!((ClearArrayNode*)n)->is_large() && n->in(2)->bottom_type()->is_long()->is_con());
+  match(Set dummy (ClearArray cnt base));
+  effect(TEMP tmp,TEMP zero,  KILL cr);
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/x86_64.ad:10860]: -> `(TEMP tmp, TEMP zero, KILL cr);`

File: src/hotspot/cpu/x86/x86_32.ad [hunk 5]
```diff
@@ -11546,6 +11546,18 @@ instruct rep_stos_large(eCXRegI cnt, eDIRegP base, regD tmp, eAXRegI zero, Unive
   ins_pipe( pipe_slow );
 %}
 
+instruct rep_stos_im(immI cnt, eRegP base, regD tmp, rRegI zero, Universe dummy, eFlagsReg cr)
+%{
+  predicate(!((ClearArrayNode*)n)->is_large() && n->in(2)->bottom_type()->is_int()->is_con());
+  match(Set dummy (ClearArray cnt base));
+  effect(TEMP tmp,TEMP zero,  KILL cr);
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/x86_32.ad:11553]: -> `effect(TEMP tmp, TEMP zero, KILL cr);`

File: src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp [hunk 6]
```diff
@@ -248,6 +248,7 @@ void MacroAssembler::copy64_avx(Register dst, Register src, Register index, XMMR
   }
 }
 
+
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86_arrayCopy_avx3.cpp:251]: Remove extra newline.

File: src/hotspot/cpu/x86/macroAssembler_x86.hpp [hunk 7]
```diff
@@ -1835,6 +1850,8 @@ class MacroAssembler: public Assembler {
   void copy64_avx(Register dst, Register src, Register index, XMMRegister xmm,
                   bool conjoint, int shift = Address::times_1, int offset = 0,
                   bool use64byteVector = false);
+
+
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.hpp:1854]: Remove extra newline.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 8]
```diff
@@ -8049,6 +8137,64 @@ void MacroAssembler::evmovdqu(BasicType type, KRegister kmask, Address dst, XMMR
...
+  evmovdqu(type[shift], mask, Address(dst, disp), xmm, Assembler::AVX_256bit);
+}
+
+
+void MacroAssembler::fill32_avx(Register dst, int disp, XMMRegister xmm) {
+  assert(MaxVectorSize >= 32, "vector length should be >= 32");
+  vmovdqu(Address(dst, disp), xmm);
+}
+
+
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:8184]: One newline between methods is sufficient (same for at other places).

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 9]
```diff
@@ -4954,34 +4958,117 @@ void MacroAssembler::xmm_clear_mem(Register base, Register cnt, XMMRegister xtmp
...
+          kmovwl(k2, rtmp);
+          evmovdqu(T_LONG, k2, Address(base, disp), xtmp, Assembler::AVX_512bit);
+        } else {
+          evmovdqu(T_LONG, k0, Address(base, disp), xtmp, Assembler::AVX_256bit);
+          movl(rtmp, 0x7);
+          kmovwl(k2, rtmp);
+          evmovdqu(T_LONG, k2, Address(base, disp + 32), xtmp, Assembler::AVX_256bit);
+        }
+        break;
+      case 6:
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:5028]: Please order case statements by increasing value.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp [hunk 10]
```diff
@@ -4927,22 +4927,26 @@ void MacroAssembler::verified_entry(int framesize, int stack_bang_size, bool fp_
...
-void MacroAssembler::xmm_clear_mem(Register base, Register cnt, XMMRegister xtmp) {
+#if COMPILER2_OR_JVMCI
+
+// clear memory of size 'cnt' qwords, starting at 'base' using XMM/YMM/ZMM registers
+void MacroAssembler::xmm_clear_mem(Register base, Register cnt, Register rtmp, XMMRegister xtmp) {
   // cnt - number of qwords (8-byte words).
   // base - start address, qword aligned.
   Label L_zero_64_bytes, L_loop, L_sloop, L_tail, L_end;
-  if (UseAVX >= 2) {
+  bool use64byteVector = MaxVectorSize == 64 && AVX3Threshold == 0;
```
Comment (TobiHartmann) [inline src/hotspot/cpu/x86/macroAssembler_x86.cpp:4937]: The comment for `AVX3Threshold` says:
```
  "Minimum array size in bytes to use AVX512 intrinsics"
  "for copy, inflate and fill. When this value is set as zero"
  "compare operations can also use AVX512 intrinsics.")
```

Should we mention clear memory there as well?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Performance numbers look good. No regression and some nice improvements (up to 13%) for some crypto microbenchmarks.

Code looks reasonable to me but assembly snippets are hard to review. I've added some style comments.

@shipilev who implemented [JDK-8146801](https://bugs.openjdk.java.net/browse/JDK-8146801) might also want to have a look.

For stability, this should wait until JDK 17 repos are forked today.

File: src/hotspot/cpu/x86/macroAssembler_x86.cpp (hunk 10 above)
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:4937]: Hi Nils, comments already cover "fill", clear memory is semantically equivalent to fill by zero.

Comment (TobiHartmann) [reply src/hotspot/cpu/x86/macroAssembler_x86.cpp:4937]: Okay, that's fine.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me but @vnkozlov should also finish his review before this is integrated.

**Review by vnkozlov — APPROVED** [review]
Passed my tier1-tier3 testing with UseAVX=3.

**Comment by jatin-bhateja** [comment]
> Passed my tier1-tier3 testing with UseAVX=3.

Thanks @vnkozlov  and @TobiHartmann

**Pull request closed without merging**